pkg/controller/daemon/manager.go

```diff
@@ -0,0 +1,475 @@
+	// more pods until all the effects (expectations) of a daemon controller's create/delete have been observed.
+	dm.podStore.Store, dm.podController = framework.NewInformer(
+		&cache.ListWatch{
+			ListFunc: func() (runtime.Object, error) {
+				return dm.kubeClient.Pods(api.NamespaceAll).List(labels.Everything(), fields.Everything())
+			},
+			WatchFunc: func(rv string) (watch.Interface, error) {
+				return dm.kubeClient.Pods(api.NamespaceAll).Watch(labels.Everything(), fields.Everything(), rv)
+			},
+		},
+		&api.Pod{},
+		FullDaemonPodResyncPeriod,
+		framework.ResourceEventHandlerFuncs{
+			AddFunc:    dm.addPod,
+			UpdateFunc: dm.updatePod,
+			DeleteFunc: dm.deletePod,
+		},
+	)
+	// Watch for new nodes or updates to nodes - daemons are launched on new nodes, and possibly when labels on nodes change,
+	dm.nodeStore.Store, dm.nodeController = framework.NewInformer(
+		&cache.ListWatch{
+			ListFunc: func() (runtime.Object, error) {
+				return dm.kubeClient.Nodes().List(labels.Everything(), fields.Everything())
+			},
+			WatchFunc: func(rv string) (watch.Interface, error) {
+				return dm.kubeClient.Nodes().Watch(labels.Everything(), fields.Everything(), rv)
+			},
+		},
+		&api.Node{},
+		FullNodeResyncPeriod,
+		framework.ResourceEventHandlerFuncs{
+			AddFunc:    dm.addNode,
+			UpdateFunc: dm.updateNode,
+			DeleteFunc: func(node interface{}) {},
+		},
+	)
+	dm.syncHandler = dm.syncDaemon
+	return dm
+}
+
+// Run begins watching and syncing daemons.
+func (dm *DaemonManager) Run(workers int, stopCh <-chan struct{}) {
+	go dm.dcController.Run(stopCh)
+	go dm.podController.Run(stopCh)
+	go dm.nodeController.Run(stopCh)
+	for i := 0; i < workers; i++ {
+		go util.Until(dm.worker, time.Second, stopCh)
+	}
+	<-stopCh
+	glog.Infof("Shutting down Daemon Controller Manager")
+	dm.queue.ShutDown()
+}
+
+func (dm *DaemonManager) worker() {
+	for {
+		func() {
+			key, quit := dm.queue.Get()
+			if quit {
+				return
+			}
+			defer dm.queue.Done(key)
+			err := dm.syncHandler(key.(string))
+			if err != nil {
+				glog.Errorf("Error syncing daemon controller with key %s: %v", key.(string), err)
+			}
+		}()
+	}
+}
+
+func (dm *DaemonManager) enqueueAllDaemons() {
+	glog.V(4).Infof("Enqueueing all daemons")
+	daemons, err := dm.dcStore.List()
+	if err != nil {
+		glog.Errorf("Error enqueueing daemon controllers: %v", err)
+		return
+	}
+	for i := range daemons {
+		dm.enqueueController(&daemons[i])
+	}
+}
+
+func (dm *DaemonManager) enqueueController(obj interface{}) {
+	key, err := controller.KeyFunc(obj)
+	if err != nil {
+		glog.Errorf("Couldn't get key for object %+v: %v", obj, err)
+		return
+	}
+	dm.queue.Add(key)
+}
+
+func (dm *DaemonManager) getPodDaemon(pod *api.Pod) *expapi.Daemon {
+	controllers, err := dm.dcStore.GetPodDaemons(pod)
+	if err != nil {
+		glog.V(4).Infof("No controllers found for pod %v, daemon manager will avoid syncing", pod.Name)
+		return nil
+	}
+	return &controllers[0]
```

Comment:
We decided while implementing the daemon controller that having multiple controllers for a daemon pod is a user error so we didn't handle this case (since non deterministic behavior is acceptable in the case of a user error). This is exactly how the replication controller worked when daemon was implemented. We could do whatever the rc does now :)

> On 8 Sep, 2015, at 2:55 pm, Mike Danese <notifications@github.com> wrote:
> 
> In pkg/controller/daemon/manager.go:
> 
> > +func (dm *DaemonManager) enqueueController(obj interface{}) {
> > +	key, err := controller.KeyFunc(obj)
> > +	if err != nil {
> > +		glog.Errorf("Couldn't get key for object %+v: %v", obj, err)
> > +		return
> > +	}
> > +	dm.queue.Add(key)
> > +}
> > +
> > +func (dm *DaemonManager) getPodDaemon(pod *api.Pod) *expapi.Daemon {
> > +	controllers, err := dm.dcStore.GetPodDaemons(pod)
> > +	if err != nil {
> > +		glog.V(4).Infof("No controllers found for pod %v, daemon manager will avoid syncing", pod.Name)
> > +		return nil
> > +	}
> > +	return &controllers[0]
> This also seems non-deterministic...
> 
> —
> Reply to this email directly or view it on GitHub.
>